### ISSUE TITLE
load crl from aws s3 for ssl connection

### ISSUE THREAD

**Issue by ackratos** [vernemq/vernemq#1448]
Hi, I saw some company manage their enterprise TLS certificates on aws acm with private ca. The private ca support revoke cert by CRL which is stored publicly on s3 (in der format). 

Is there any plan vernemq would support setting `listener.ssl.crlfile` in config file to an HTTP endpoint and fetch it periodically (maybe make the interval and format of CRL file can be configured as well)?

**Comment by ioolkos** [comment]
Hi @ackratos thanks, this seems like a good idea to me. (PRs welcome...)

I would like to open a discussion here about the handling of CRLs in VerneMQ. How have people been using/implementing this? what seems best practice? what is best practice but possibly not implemented?

**Comment by ackratos** [comment]
> I would like to open a discussion here about the handling of CRLs in VerneMQ. How have people been using/implementing this? what seems best practice? what is the best practice but possibly not implemented?

Hi, thanks for your kind reply. I am also interested in this topic:) I know nothing about erlang, but if this is the use case for many people I'd like to contribute, I think the implementation should be simple (maybe I can copy/paste the logic for vmq_passwd file reloading and setting logic). The problem seems not too much use case or guidance on how to deploy vernemq on cloud flatform like aws with secure setting as aws iot. I even don't see this crl option when I was scanned through the document

**Comment by codeadict** [comment]
Erlang itself supports grabbing crls from a file over HTTP since version 19, we may need to improve the configuration options for that. Basically two settings need to be enabled via `ssl_options`: `{crl_check, true}` which instructs to check for crls and  `{crl_cache, {ssl_crl_cache, {internal, [{http, 5000}]}}}` that means that crl cache will be refreshed over http with a 5000 timeout. The urls to download crls needs to be specified in the `cRLDistributionPoints` x509 extension of the Server CA certificate, for example using openssl config would be similar to:

```conf
[ca]
x509_extensions   = user_cert

[user_cert]
crlDistributionPoints=crl_section

[crl_section]
fullname=URI:http://somehost.com/crl.pem
```

Check http://erlang.org/doc/man/ssl.html#type-crl_cache_opts for better understanding, we may need to document this better and test it, probably improve the config options in VerneMQ for it.

**Comment by ioolkos** [comment]
Thanks @codeadict.
@eduardogmisiuk how do you handle CRLs in your setup?

**Comment by ioolkos** [comment]
So, as @codeadict  mentioned, `ssl_crl_cache` is an internal CRL cache module that can be populated actively (by calling the correct insert function), or automatically loaded with a HTTP call.
You can't configure the HTTP endpoint, I think, Erlang doc says: 
 ```
 "Enables fetching of CRLs specified as http URIs inX509 certificate extensions."

**Comment by eduardogmisiuk** [comment]
@ioolkos we set the CRL file in VerneMQ config as `listener.ssl.default.crlfile =  /vernemq/cert/ca.crl` and update it using a Shell script in Cron.

**Comment by codeadict** [comment]
Yeah, @eduardogmisiuk approach is to grab it from the local filesystem. Erlang's approach for http works but the way one has to specify the URL is a bit unfriendly to the user, not difficult tho. Maybe good documentation can fix it in the meantime and they have a behaviour one can implement and maybe standarize the setting like `listener.ssl.default.crlfile = http://somehost.com/crl.pem`. Anyways, meantime better docs can allow users to achieve this and longer time a custom implementation to populate the `ssl_crl_cache`

**Comment by ioolkos** [comment]
Maybe interesting: https://grafana.com/blog/2020/03/04/pro-tip-how-to-monitor-client-certificate-expirations-with-prometheus/

**Comment by onrao** [comment]
we are looking for this feature support. @ioolkos  and @codeadict  requesting to enable this configuration through docker environment variables as well for containerized usage.

**Comment by ioolkos** [comment]
@onrao thanks for "requesting" a feature :)
While this is 100% welcome, it helps to be specific. (I guess you mean grabbing CRLS over HTTP).
It also helps to indicate in what way you can help from your side to make it happen. 
As mentioned above, we're open to PRs any time.
